File: DigitalIcebreakers.EndToEndTests/Helpers/BrowserFactory.cs
using System;
using PlaywrightSharp;
using PlaywrightSharp.Chromium;
using System.Threading.Tasks;
using System.Linq;

namespace DigitalIcebreakers.EndToEndTests
{
    public class BrowserFactory
    {
        private readonly IServiceProvider _provider;
        private readonly TestSettings _settings;
        private readonly IPlaywright _playwright;
        private readonly DisposableServices _disposableServices;

        public BrowserFactory(IServiceProvider provider, TestSettings settings, IPlaywright playwright, DisposableServices disposableServices)
        {
            _provider = provider;
            _settings = settings;
            _playwright = playwright;
            _disposableServices = disposableServices;
        }

        private async Task<IBrowser> Create(bool? headless = null)
        {
            IBrowser browser = null;
            switch (_settings.Browser)
            {
                case "Chrome":
                    {
                        browser = await _playwright.Chromium.LaunchAsync(headless: headless ?? _settings.Headless);
                        break;
                    }
                case "Firefox":
                    {
                        browser = await _playwright.Firefox.LaunchAsync(headless: headless ?? _settings.Headless);
                        break;
                    }
                case "Webkit":
                    {
                        browser = await _playwright.Webkit.LaunchAsync(headless: headless ?? _settings.Headless);
                        break;
                    }
                default: throw new Exception("Unknown browser");
            }

            _disposableServices.ServicesAsync.Add(browser);
            return browser;
        }

        public async Task<Presenter> CreatePresenter()
        {
            var browser = await Create();
            var page = await browser.NewPageAsync();
            await page.GoToAsync(_settings.Url);
            await page.ClickAsync("text='Present'", delay: 1000);
            await page.ClickAsync("text='Create'", timeout: 2000);
            var url = await page.GetAttributeAsync("[data-testid=qrcode-link]", "href", 1000);
            return new Presenter(browser, page, url);
        }

        public async Task<Player> CreatePlayer(string lobbyUrl, string playerName = "test-user", bool? headless = null)
        {
            var browser = await Create(headless);
            var page = await browser.NewPageAsync();
            await JoinLobby(lobbyUrl, page, playerName);
            return new Player(browser, page);
        }

        public async Task<Player[]> CreatePlayers(string lobbyUrl, int playerCount)
        {
            return await Task.WhenAll(Enumerable.Range(1, playerCount)
                .ToList()
                .Select(ix => CreatePlayer(lobbyUrl, $"Player {ix}")));
        }

        public static async Task JoinLobby(string lobbyUrl, IPage page, string playerName = "test-user")
        {
            await page.GoToAsync(lobbyUrl);
            await page.TypeAsync("[id=user-name]", playerName);
            var joinButton = await page.GetByTestId("join-lobby-button");
            await joinButton.ClickAsync();
        }

        public async Task<Player> CreatePlayerByJoinCode(string joinCode, string playerName = "test-user", bool? headless = null)
        {
            var browser = await Create(headless);
            var page = await browser.NewPageAsync();
            await page.GoToAsync(_settings.Url);
            await page.ClickAsync("text='Join'", delay: 1000);
            await page.TypeAsync("[id=lobby-code]", joinCode);
            await page.ClickByTestId("join-lobby");
            await page.TypeAsync("[id=user-name]", playerName);
            await page.ClickByTestId("join-lobby-button");
            return new Player(browser, page);
        }
    }
}

File: web/e2e/helpers/browserFactory.ts
import { Browser, BrowserContext, Page } from "@playwright/test";
import { Presenter } from "./presenter";
import { Player } from "./player";

export class BrowserFactory {
  private baseURL: string;
  private contexts: BrowserContext[] = [];

  constructor(baseURL?: string) {
    this.baseURL = baseURL || process.env.BASE_URL || "http://localhost:5273";
  }

  private async newContext(browser: Browser): Promise<BrowserContext> {
    const context = await browser.newContext();
    this.contexts.push(context);
    return context;
  }

  async createPresenter(browser: Browser): Promise<Presenter> {
    const context = await this.newContext(browser);
    const page = await context.newPage();

    await page.goto(this.baseURL);
    await this.waitForConnected(page);
    await page.getByRole("button", { name: "Present" }).click();
    await page.getByRole("button", { name: "Create" }).click();

    const url = await page
      .getByTestId("qrcode-link")
      .first()
      .getAttribute("href");
    if (!url) {
      throw new Error("Could not get lobby URL from QR code link");
    }

    return new Presenter(browser, page, url);
  }

  async createPlayer(
    browser: Browser,
    lobbyUrl: string,
    playerName: string = "test-user"
  ): Promise<Player> {
    const context = await this.newContext(browser);
    const page = await context.newPage();

    await this.joinLobby(lobbyUrl, page, playerName);
    return new Player(browser, page);
  }

  async createPlayers(
    browser: Browser,
    lobbyUrl: string,
    playerCount: number
  ): Promise<Player[]> {
    const playerPromises = Array.from({ length: playerCount }, (_, i) =>
      this.createPlayer(browser, lobbyUrl, `Player ${i + 1}`)
    );

    return Promise.all(playerPromises);
  }

  public async joinLobby(
    lobbyUrl: string,
    page: Page,
    playerName: string = "test-user"
  ): Promise<void> {
    await page.goto(lobbyUrl);
    // sends made before the transport connection is up are silently dropped
    await this.waitForConnected(page);
    // target the register form's name field specifically: the lobby route
    // briefly shows the join-code view, whose code input would otherwise
    // swallow a generic textbox fill before the register view mounts
    const textBox = page.locator("#user-name");
    await textBox.waitFor({ state: "visible" });
    await textBox.fill(playerName);
    await page.getByTestId("join-lobby-button").click();
    // joining is only complete once the register form unmounts; interacting
    // earlier can target inputs that are about to be replaced by the game view
    await page
      .getByTestId("join-lobby-button")
      .waitFor({ state: "detached" });
  }

  private async waitForConnected(page: Page): Promise<void> {
    await page
      .locator('[data-testid="connection-status"][data-status="Connected"]')
      .first()
      .waitFor({ state: "attached" });
  }

  async createPlayerByJoinCode(
    browser: Browser,
    joinCode: string,
    playerName: string = "test-user"
  ): Promise<Player> {
    const context = await this.newContext(browser);
    const page = await context.newPage();

    await page.goto(this.baseURL);
    await page.getByRole("button", { name: "Join" }).click();
    await page.fill("#lobby-code", joinCode);
    await page.getByTestId("join-lobby").click();
    await page.fill("#user-name", playerName);
    await page.getByTestId("join-lobby-button").click();

    return new Player(browser, page);
  }

  async cleanup(): Promise<void> {
    await Promise.all(this.contexts.map((context) => context.close()));
    this.contexts = [];
  }

  static async joinLobby(
    lobbyUrl: string,
    page: Page,
    playerName: string = "test-user"
  ): Promise<void> {
    await page.goto(lobbyUrl);
    await page.fill("#user-name", playerName);
    await page.getByTestId("join-lobby-button").click();
  }
}
